File: editor/src/de/fatox/meta/shader/MetaShaderLibrary.java
package de.fatox.meta.shader;

import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.files.FileHandle;
import com.badlogic.gdx.utils.Array;
import com.badlogic.gdx.utils.Json;
import com.badlogic.gdx.utils.ObjectMap;
import de.fatox.meta.Meta;
import de.fatox.meta.api.dao.GLShaderData;
import de.fatox.meta.api.graphics.GLShaderHandle;
import de.fatox.meta.api.graphics.MetaGLShader;
import de.fatox.meta.api.ui.UIManager;
import de.fatox.meta.ide.ProjectManager;
import de.fatox.meta.injection.Inject;
import de.fatox.meta.injection.Singleton;

@Singleton
public class MetaShaderLibrary {
    private static final String META_SHADER_SUFFIX = ".msh";
    private static final String INTERNAL_SHADER_PATH = "meta/shaders";
    @Inject
    private ProjectManager projectManager;
    @Inject
    private Json json;
    @Inject
    private UIManager uiManager;

    private ObjectMap<String, GLShaderHandle> loadedShaders = new ObjectMap<>();
    private Array<GLShaderHandle> metaShaders = new Array<>();

    public MetaShaderLibrary() {
        Meta.inject(this);
        projectManager.addOnLoadListener((evt) -> {
            loadProjectShaders();
            return false;
        });
        Gdx.app.postRunnable(this::loadProjectShaders);
    }

    public GLShaderHandle loadShader(FileHandle shaderHandle) {
        GLShaderData shaderData = json.fromJson(GLShaderData.class, shaderHandle.readString());
        FileHandle projRoot = projectManager.getCurrentProjectRoot();
        FileHandle vertHandle = projRoot.child(shaderData.getVertexFilePath());
        FileHandle fragHandle = projRoot.child(shaderData.getFragmentFilePath());
        if (vertHandle.exists() && !vertHandle.isDirectory() && fragHandle.exists() && !fragHandle.isDirectory()) {
            GLShaderHandle handle = new GLShaderHandle(shaderHandle, vertHandle, fragHandle, shaderData);
            metaShaders.add(handle);
            loadedShaders.put(projectManager.relativize(shaderHandle), handle);
            return handle;
        }
        return null;
    }

    public GLShaderHandle newShader(GLShaderData data) {
        FileHandle newShaderHandle = projectManager.save(INTERNAL_SHADER_PATH + data.getName() + META_SHADER_SUFFIX, data);
        return loadShader(newShaderHandle);
    }


    public Array<GLShaderHandle> getLoadedShaders() {
        return metaShaders;
    }

    public MetaGLShader getDefaultShaderPath(GLShaderHandle shaderHandle) {
        return null;
    }

    public GLShaderHandle getShaderHandle(String metaShaderPath) {
        if (loadedShaders.containsKey(metaShaderPath)) {
            return loadedShaders.get(metaShaderPath);
        }
        return null;
    }

    public void loadProjectShaders() {
        if (projectManager.getCurrentProject() != null) {
            FileHandle shaderFolder = projectManager.getCurrentProjectRoot().child(INTERNAL_SHADER_PATH);
            if (shaderFolder.exists()) {
                for (FileHandle metaShaderDef : shaderFolder.list(pathname -> pathname.getName().endsWith(META_SHADER_SUFFIX))) {
                    loadShader(metaShaderDef);
                }
            }
        }
    }

    public String getDefaultShaderPath() {
        return projectManager.relativize(loadedShaders.values().next().getShaderHandle());
    }
}

File: editor/src/de/fatox/meta/shader/MetaShaderLibrary.kt
package de.fatox.meta.shader

import com.badlogic.gdx.Gdx
import com.badlogic.gdx.files.FileHandle
import com.badlogic.gdx.utils.Array
import com.badlogic.gdx.utils.Json
import com.badlogic.gdx.utils.ObjectMap
import de.fatox.meta.api.graphics.GLShaderHandle
import de.fatox.meta.api.model.GLShaderData
import de.fatox.meta.ide.ProjectManager
import de.fatox.meta.injection.MetaInject.Companion.lazyInject
import de.fatox.meta.injection.Singleton
import de.fatox.meta.listener.MetaNotifier

@Singleton
class MetaShaderLibrary : MetaNotifier() {
	private val projectManager: ProjectManager by lazyInject()
	private val json: Json by lazyInject()

	private val loadedShaders = ObjectMap<String, GLShaderHandle>()
	private val metaShaders = Array<GLShaderHandle>()

	val defaultShaderPath: String
		get() = projectManager.relativize(loadedShaders.values().next().shaderHandle)

	init {

		projectManager.addOnLoadListener {
			loadedShaders.clear()
			metaShaders.clear()
			loadDefaultShader()
			loadProjectShaders()
			false
		}
		Gdx.app.postRunnable { this.loadProjectShaders() }
	}

	private fun loadDefaultShader() {
		val internal = Gdx.files.internal("shaders/Default.msh")
		val loadShader = loadShader(internal, false)
		loadedShaders.put("shaders/Default.msh", loadShader)
		notifyListeners()
	}

	fun loadShader(shaderHandle: FileHandle, projectFile: Boolean): GLShaderHandle? {
		val shaderData = json.fromJson(GLShaderData::class.java, shaderHandle.readString())
		val projRoot = projectManager.currentProjectRoot
		val vertHandle = projRoot.child(shaderData.vertexFilePath)
		val fragHandle = projRoot.child(shaderData.fragmentFilePath)
		if (vertHandle.exists() && !vertHandle.isDirectory && fragHandle.exists() && !fragHandle.isDirectory) {
			val handle = GLShaderHandle(shaderHandle, vertHandle, fragHandle, shaderData)
			metaShaders.add(handle)
			loadedShaders.put(if (projectFile) projectManager.relativize(shaderHandle) else shaderHandle.path(), handle)
			return handle
		}
		return null
	}

	fun newShader(data: GLShaderData): GLShaderHandle? {
		val newShaderHandle = projectManager.save(INTERNAL_SHADER_PATH + "/" + data.name + META_SHADER_SUFFIX, data)
		val loadShader = loadShader(newShaderHandle, true)
		notifyListeners()
		return loadShader
	}


	fun getLoadedShaders(): Array<GLShaderHandle> {
		return metaShaders
	}

	fun getFirstShader(): GLShaderHandle {
		return metaShaders.first()
	}

	fun getShaderHandle(metaShaderPath: String): GLShaderHandle? {
		return if (loadedShaders.containsKey(metaShaderPath)) {
			loadedShaders.get(metaShaderPath)
		} else null
	}

	fun loadProjectShaders() {
		if (projectManager.currentProject != null) {
			val shaderFolder = projectManager.currentProjectRoot.child(INTERNAL_SHADER_PATH)
			if (shaderFolder.exists()) {
				for (metaShaderDef in shaderFolder.list { pathname -> pathname.name.endsWith(META_SHADER_SUFFIX) }) {
					loadShader(metaShaderDef, true)
				}
			}
		}
		notifyListeners()
	}

	companion object {
		private const val META_SHADER_SUFFIX = ".msh"
		private const val INTERNAL_SHADER_PATH = "meta/shaders"
	}
}
